### PR TITLE
Implement random partitioner

### PR DESCRIPTION
**Pull request by id**

### CODE REVIEW

File: web/controllers/api/v1/message_controller.ex [hunk 1]
```diff
@@ -6,26 +6,58 @@ defmodule Kastlex.API.V1.MessageController do
...
+      {:producer_not_found, _topic} ->
+        send_json(conn, 404, %{error: "Unknown topic"})
+      {:producer_not_found, _topic, _partition} ->
+        send_json(conn, 404, %{error: "Unknown partition"})
+      reason ->
+        Logger.error "Failed to produce to #{topic}: #{reason}"
+        send_json(conn, 503, %{error: "Service unavailable"})
+    end
+  end
+  defp try_produce(conn, topic, [p | partitions], key, value, _last_error) do
```
Comment (zmstone) [inline web/controllers/api/v1/message_controller.ex]: add a clause to match last_error, and only try next partition when it's NOT UnknownTopicOrPartition producer_not_found.

Comment (zmstone) [inline web/controllers/api/v1/message_controller.ex]: or perhaps only retry when it's not leader.

File: web/controllers/api/v1/message_controller.ex [hunk 2]
```diff
@@ -6,26 +6,58 @@ defmodule Kastlex.API.V1.MessageController do
...
+
+  # no partition, use random partitioner
+  def produce(conn, %{"topic" => topic} = params) do
+    key = Map.get(params, "key", "")
+    {:ok, value, conn} = read_body(conn)
+    case :brod_client.get_partitions_count(:kastlex, topic) do
+      {:ok, partitions_cnt} ->
+        partitions = :lists.seq(0, partitions_cnt - 1) |> Enum.shuffle
+        try_produce(conn, topic, partitions, key, value)
       {:error, :UnknownTopicOrPartition} ->
```
Comment (zmstone) [inline web/controllers/api/v1/message_controller.ex]: error reason -> http code mapping can be abstracted into a common function share with try_produce

File: web/controllers/api/v1/message_controller.ex (hunk 1 above)
Comment (id) [inline web/controllers/api/v1/message_controller.ex]: UnknownTopicOrPartition should be caught by brod_client:get_partitions_count, or?

Comment (zmstone) [inline web/controllers/api/v1/message_controller.ex]: not if the partition is given in request parameters.

Comment (id) [inline web/controllers/api/v1/message_controller.ex]: In this case we don't have any more partitions to retry.

File: web/controllers/api/v1/message_controller.ex [hunk 3]
```diff
@@ -6,26 +6,58 @@ defmodule Kastlex.API.V1.MessageController do
...
 
-  # TODO: partitioner
   def produce(conn, %{"topic" => topic, "partition" => partition} = params) do
     {partition, _} = Integer.parse(partition)
     key = Map.get(params, "key", "")
     {:ok, value, conn} = read_body(conn)
-    case :brod.produce_sync(:kastlex, topic, partition, key, value) do
-      :ok ->
-        send_resp(conn, 204, "")
+    try_produce(conn, topic, partition, key, value)
```
Comment (zmstone) [inline web/controllers/api/v1/message_controller.ex]: call [parition] here to save a function clause for try_produce/5

File: web/controllers/api/v1/message_controller.ex (hunk 1 above)
Comment (zmstone) [inline web/controllers/api/v1/message_controller.ex]: not sure how that is related. isn't the error from the last tried ?
e.g. if there is only one partition in a topic, and the user request to produce to partition 1,
we'll still end up with unknown partition, or ?

Comment (zmstone) [inline web/controllers/api/v1/message_controller.ex]: get_partitions_count is only called when partition is missing from the request parameters.

Comment (id) [inline web/controllers/api/v1/message_controller.ex]: Never mind, I'll retry only on NotLeaderForPartition.

On a side note, doesn't brod return producer_not_found only when auto_start_producers is false?

Comment (zmstone) [inline web/controllers/api/v1/message_controller.ex]: when auto_start_producers is enabled, `produce` call will make `brod_client` try to start the producer, before starting the topic sup,`brod_client` check existence of the topic (and partition), return `{error, ?EC_UNKNOWN_TOPIC_OR_PARTITION}` if not found.

Comment (id) [inline web/controllers/api/v1/message_controller.ex]: Yes, that's what I mean.

File: web/controllers/api/v1/message_controller.ex [hunk 4]
```diff
@@ -6,23 +6,59 @@ defmodule Kastlex.API.V1.MessageController do
...
+
+  # no partition, use random partitioner
+  def produce(conn, %{"topic" => topic} = params) do
+    key = Map.get(params, "key", "")
+    {:ok, value, conn} = read_body(conn)
+    case :brod_client.get_partitions_count(:kastlex, topic) do
+      {:ok, partitions_cnt} ->
+        partitions = :lists.seq(0, partitions_cnt - 1) |> Enum.shuffle
+        try_produce(conn, topic, partitions, key, value, nil)
+      {:error, :UnknownTopicOrPartition} = error ->
```
Comment (id) [inline web/controllers/api/v1/message_controller.ex]: set of possible errors from :brod_client.get_partitions_count/2 and :brod.produce_sync/5 is different, does not make much sense to abstract error handling code for them in one common function

**Pull request merged**